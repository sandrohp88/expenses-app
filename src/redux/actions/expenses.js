import expensesRef from '../../firebase/firebase'
// ADD_EXPENSE
const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
})

// ADD_EXPENSE_ASYNC
const addExpenseAsync = (expenseData = {}) => {
  return async dispatch => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }
    const response = await expensesRef.add(expense)
    return dispatch(addExpense({ id: response.id, ...expense }))
  }
}

// REMOVE_EXPENSE
const removeExpense = (id = '') => ({ type: 'REMOVE_EXPENSE', id })

// EDIT_EXPENSE
const editExpense = (id, updates) => ({ type: 'EDIT_EXPENSE', id, updates })

// SET_EXPENSES
const setExpenses = expenses => ({ type: 'SET_EXPENSES', expenses })

// GET_EXPENSES_ASYNC
const getExpensesAsync = () => {
  return async dispatch => {
    const response = await expensesRef.get()
    const expenses = []
    response.forEach(expense => {
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
      } = expense.data()
      expenses.push({id:expense.id, description, note, amount, createdAt })
    })
    return dispatch(setExpenses(expenses))
  }
}

export {
  getExpensesAsync,
  addExpense,
  addExpenseAsync,
  removeExpense,
  editExpense,
  setExpenses
}
