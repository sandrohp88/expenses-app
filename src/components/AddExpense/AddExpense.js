import React from 'react'
import { connect } from 'react-redux'
import { ExpenseForm } from '../ExpenseForm/ExpenseForm'
import { addExpense } from '../../redux/actions/expenses'
export class AddExpense extends React.Component {
  onSubmit = expense => {
    this.props.addExpense(expense)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <h1>Add expense component</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}
const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
})
export default connect(
  undefined,
  mapDispatchToProps
)(AddExpense)
