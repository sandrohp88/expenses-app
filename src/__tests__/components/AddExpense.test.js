import React from 'react'
import { shallow } from 'enzyme'
import { AddExpense } from '../../components/AddExpense/AddExpense'
import expenses from '../../sampleData/expenses'

let addExpense, history, wrapper
beforeEach(() => {
  addExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<AddExpense addExpense={addExpense} history={history} />)
})

test('render Add page without crash', () => {
  expect(wrapper).toMatchSnapshot()
})
test('handle onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(addExpense).toHaveBeenLastCalledWith(expenses[2])
})
