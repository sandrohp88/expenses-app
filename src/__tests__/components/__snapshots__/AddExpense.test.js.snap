// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Add page without crash 1`] = `
<div>
  <h1>
    Add expense component
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
